### DISCUSSION TITLE
Issues building 4.6.4/4.6.6 on Ubuntu 23.10 - error: ‘QWaylandApplication’ is not a member of ‘QNativeInterface’

### DISCUSSION THREAD

**Discussion by Bro-Account** [martinrotter/rssguard#1338] [Q&A] [answered]
Hi,

Unfortunately I'm having issues again building the new release on Ubuntu (RPi). Any ideas how I can fix that?

> $ cat /etc/os-release 
> PRETTY_NAME="Ubuntu 23.10"
> NAME="Ubuntu"
> VERSION_ID="23.10"
> VERSION="23.10 (Mantic Minotaur)"
> VERSION_CODENAME=mantic
> ID=ubuntu
> ID_LIKE=debian
> HOME_URL="https://www.ubuntu.com/"
> SUPPORT_URL="https://help.ubuntu.com/"
> BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
> PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
> UBUNTU_CODENAME=mantic
> LOGO=ubuntu-logo

>  $uname -a
> Linux pibuntu 6.5.0-1012-raspi #15-Ubuntu SMP PREEMPT_DYNAMIC Thu Feb 22 15:23:50 UTC 2024 aarch64 aarch64 aarch64 GNU/Linux
> 

> [ 96%] Building CXX object src/librssguard/CMakeFiles/rssguard.dir/gui/mediaplayer/libmpv/libmpvwidget.cpp.o
> /home/bro/Documents/rssguard/rssguard-4.6.6/src/librssguard/gui/mediaplayer/libmpv/libmpvwidget.cpp: In member function ‘virtual void LibMpvWidget::initializeGL()’:
> /home/bro/Documents/rssguard/rssguard-4.6.6/src/librssguard/gui/mediaplayer/libmpv/libmpvwidget.cpp:120:60: error: ‘QWaylandApplication’ is not a member of ‘QNativeInterface’; did you mean ‘QX11Application’?
>   120 |     display.data = qApp->nativeInterface<QNativeInterface::QWaylandApplication>()->display();
>       |                                                            ^~~~~~~~~~~~~~~~~~~
>       |                                                            QX11Application
> /home/bro/Documents/rssguard/rssguard-4.6.6/src/librssguard/gui/mediaplayer/libmpv/libmpvwidget.cpp:120:80: error: no matching function for call to ‘Application::nativeInterface<<expression error> >()’
>   120 |     display.data = qApp->nativeInterface<QNativeInterface::QWaylandApplication>()->display();
> In file included from /usr/include/aarch64-linux-gnu/qt6/QtCore/qcoreapplication.h:16,
>                  from /usr/include/aarch64-linux-gnu/qt6/QtCore/QCoreApplication:1,
>                  from /home/bro/Documents/rssguard/rssguard-4.6.6/src/librssguard/miscellaneous/iofactory.h:6,
>                  from /home/bro/Documents/rssguard/rssguard-4.6.6/src/librssguard/miscellaneous/application.h:11,
>                  from /home/bro/Documents/rssguard/rssguard-4.6.6/src/librssguard/gui/mediaplayer/libmpv/libmpvwidget.cpp:5:
> /usr/include/aarch64-linux-gnu/qt6/QtWidgets/qapplication.h:121:5: note: candidate: ‘template<class NativeInterface, class TypeInfo, class BaseType, typename std::enable_if<TypeInfo::isCompatibleWith<QApplication>, bool>::type <anonymous> > NativeInterface* QApplication::nativeInterface() const’
>   121 |     QT_DECLARE_NATIVE_INTERFACE_ACCESSOR(QApplication)
>       |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> /usr/include/aarch64-linux-gnu/qt6/QtWidgets/qapplication.h:121:5: note:   template argument deduction/substitution failed:
> /home/bro/Documents/rssguard/rssguard-4.6.6/src/librssguard/gui/mediaplayer/libmpv/libmpvwidget.cpp:120:80: error: template argument 1 is invalid
>   120 |     display.data = qApp->nativeInterface<QNativeInterface::QWaylandApplication>()->display();
> gmake[2]: *** [src/librssguard/CMakeFiles/rssguard.dir/build.make:3681: src/librssguard/CMakeFiles/rssguard.dir/gui/mediaplayer/libmpv/libmpvwidget.cpp.o] Error 1
> gmake[1]: *** [CMakeFiles/Makefile2:156: src/librssguard/CMakeFiles/rssguard.dir/all] Error 2
> gmake: *** [Makefile:136: all] Error 2

**Comment by martinrotter** [comment]
No idea sorry, it builds fine on Ubuntu github-actions tho.

**Reply by Bro-Account** [reply]
According to this bug report, this should be related to a missing package xcb-util-cursor https://github.com/xiaoyifang/goldendict-ng/issues/1471
However I'm unable to install the package.

$ sudo apt-get install xcb-util-cursor
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
E: Unable to locate package xcb-util-cursor

**Reply by shenlebantongying** [reply, marked as answer]
The rood cause is that Qt can be built without either Wayland or xcb support (The GitHub actions CI working because the one in Ubuntu has both enabled.).

These lines need to guard against these situations with 

```
#ifdef QT_FEATURE_xcb
#ifdef QT_FEATURE_wayland
```

https://github.com/martinrotter/rssguard/blob/80581489ea768b1a317c1dac10e5b96d2df0fa73/src/librssguard/gui/mediaplayer/libmpv/libmpvwidget.cpp#L113-L121

Reference: `/usr/include/qt6/QtGui/qtgui-config.h`
Related https://github.com/xiaoyifang/goldendict-ng/issues/1471#issuecomment-2249067379

**Reply by Bro-Account** [reply]
Thank you so much, I did not quite understand what I had to do, but with ChatGPT's help I was able to fix the issue and build 4.7.3. @martinrotter Can you add this to your branch?

> To fix the build process for RSSGuard considering the issue with Qt's Wayland and xcb support, you need to modify the libmpvwidget.cpp file to handle cases where these features might not be available. You need to add preprocessor directives to conditionally compile the code based on whether these features are enabled in your Qt build. Here’s how you can do it:
> 
> Open the file libmpvwidget.cpp located at src/librssguard/gui/mediaplayer/libmpv/.
> 
> Add the #ifdef preprocessor directives around the relevant code sections. The modified lines 113 to 121 should look something like this:
> 
> ```
> #ifdef QT_FEATURE_xcb
> if (QGuiApplication::platformName() == QStringLiteral("xcb")) {
>     display.type = MPV_RENDER_PARAM_X11_DISPLAY;
>     display.data = qApp->nativeInterface<QNativeInterface::QX11Application>()->display();
> }
> #endif
> 
> #ifdef QT_FEATURE_wayland
> if (QGuiApplication::platformName() == QStringLiteral("wayland")) {
>     display.type = MPV_RENDER_PARAM_WL_DISPLAY;
>     display.data = qApp->nativeInterface<QNativeInterface::QWaylandApplication>()->display();
> }
> #endif
> ```
> Save the file and re-run the build process.
> These changes ensure that the xcb and wayland specific code blocks are only included if the respective Qt features are enabled. This should help in avoiding build-time errors related to missing Qt features.

**Comment by OldCoder** [comment]
<div align="justify">
I've received email directly or indirectly from "Bro-Account" about the issue. To respond:
<br /><br />

N1. "xcb-util-cursor" may be the issue, but there are other possible explanations.

N2. The distro-level package name varies from distro to distro. For Debian distros, including Ubuntu 23.10, try the following commands:

sudo apt-get update
sudo apt-get install libxcb-cursor0
sudo apt-get install libxcb-cursor-dev

Ideally, at least one of the "install" commands should work.
</div>

**Comment by martinrotter** [comment]
OK.

683fa898e
